[mlir] Remove the mlir-spirv-cpu-runner (move to mlir-cpu-runner)

andfau-amd opened the pull request:
This commit builds on and completes the work done in 9f6c632ecda08bfff76b798c46d5d7cfde57b5e9 to eliminate the need for a separate mlir-spirv-cpu-runner binary. Since the MLIR processing is already done outside this runner, the only real difference between it and the mlir-cpu-runner is the final linking step between the nested LLVM IR modules. By moving this step into mlir-cpu-runner behind a new command-line flag (`--link-nested-modules`), this commit is able to completely remove the runner component of the mlir-spirv-cpu-runner.

The runtime libraries and the tests are moved and renamed to fit into the Execution Engine and Integration tests, following the model of the similar migration done for the CUDA Runner in D97463.

joker-eph commented:
Nice! It's great to see the progress on this, thanks for pushing this forward :)

andfau-amd commented:
> Oh, should I move the "wrappers" to the execution engine? Is that just a renaming process?

Answering myself here: yes I should, as outlined in https://github.com/llvm/llvm-project/issues/73457, and yes it is, if https://reviews.llvm.org/D97463 is the clearest example of the pattern required. I do not have the clarity of mind needed right at this moment to execute it, but I should have this done tomorrow. I think it should be straightforward.

joker-eph commented on mlir/tools/mlir-cpu-runner/mlir-cpu-runner.cpp:
```diff
@@ -17,10 +17,65 @@
+    "link-nested-modules",
+    llvm::cl::desc("Link two nested MLIR modules into a single LLVM IR module. "
+                   "Useful if both the host and device code can be run on the "
+                   "same CPU, as in mlir-spirv-cpu-runner tests."));
```
Isn't this something that could be done as a mlir-opt pass ahead of time instead of bundling "transformation logic" in the CPU runner?

kuhar replied:
+1

andfau-amd replied:
Maybe theoretically? But the problem is that mlir-opt produces MLIR, and this linking is done at the LLVM IR level, not at the MLIR one. So we would have to move the conversion to LLVM IR into mlir-opt, then the linking, then it would have to turn that back into MLIR of some kind. I'm not sure what the benefit would be. It's a lot more straightforward to leave this in the runner.

joker-eph replied:
Oh I see what you mean, sorry I overlooked the linking logic part.

This flow isn't super clear to me actually, is there still any spirv involved?
Shouldn't we use the same target offloading mechanism in all cases?

andfau-amd replied:
Yes that's a possibility, we could serialise the device code as LLVM bitcode the same way we do for e.g. SPIR-V. It's probably an unnecessary complication though. I'm willing to something like this out but maybe in a follow-up commit, as there's a few possibilities for how we could proceed?

andfau-amd replied:
Hmm, I feel like I'm pulled in two different directions here. My goal was to get rid of the separate SPIR-V CPU runner, and the most straightforward way to accomplish that has been to move the codegen steps (SPIR-V -> LLVM MLIR and LLVM MLIR -> LLVM IR) into mlir-opt and mlir-cpu-runner respectively, and that _is_ bringing it more into alignment with the other runners. But for the Vulkan SPIR-V runner, these codegen steps necessarily happen inside the runner, or rather something like them happens inside the Vulkan driver that'll be dynamically linked in (SPIR-V -> [something proprietary] -> LLVM IR -> [native binary]). So if we want the SPIR-V CPU runner to work the same as the Vulkan SPIR-V runner, then we would probably have to move the codegen back into the runner, so that the mlir-opt output can just be SPIR-V.

andfau-amd replied:
(It's relevant because this is basically just a practice run for getting rid of the Vulkan SPIR-V runner.)

rengolin replied:
We have a similar tug-of-war between our `opt` and `runner` apps (based on the same principles). 

We had to solve the problem of ingress: which IR stage (dialects, passes) to accept. `opt` takes in anything, but `runner` is more complicated, because it needs to lower to LLVM/SPIRV to actually execute, and if the IR is in `linalg`, then it's not going to work. The solution for this is to pass it through `opt` until it's in the right shape, then pipe to `runner`, but defining what the "right shape" is, is the point here.

Let's assume `runner` can only handle LLVM and SPIRV, then the main problem with merging the runners is to document which of these two can be used for which "back-end", and how to drive the back-end (tools, LLVM codegen, dispatch, JIT) with these dialects.

Regardless, I think this is super important and needs to happen. The main difference in code is which MLIR/LLVM functions are called, and some configuration issues, which we can probable common up for all these cases.

andfau-amd replied:
For my part I think this commit is still an improvement as it stands and I'd rather it be merged more or less as-is (modulo some naming changes etc), then further changes to the shape of this can be experimented with later if necessary. The SPIR-V CPU runner is pretty obscure and rarely used though, so it's not that important that it have the ideal shape, if that makes sense. If there's something worth arguing about it might be how the Vulkan SPIR-V runner changes look.

andfau-amd replied:
Like, a TODO comment? I guess I can put it in this file next to the flag/pass?

kuhar replied:
Yes, and this TODO comment can link to a github issue

joker-eph replied:
That's fine as a stepping stone, thanks again @andfau-amd for the work here.

@fabianmcg : thought about how to make this converge with the rest of the offloading infra? (or maybe it's just not a good fit?)

fabianmcg replied:
I'll take a closer look tomorrow, but at first glance, it should be possible to generalize and extract the infra from GPU and have something like a general `module-to-offload` for handling nested module compilation. The question is whether people are interested in such a thing outside GPU.

Also, as a reminder for GPUs, we already have integration https://github.com/llvm/llvm-project/blob/main/mlir/test/Integration/GPU/SYCL/gpu-reluf32-to-spirv.mlir

kuhar replied:
@fabianmcg keep in mind that while this is a spirv runner, the serialization format for the device code is also llvm. Hence the link step.

kuhar commented on mlir/docs/SPIRVToLLVMDialectConversion.md:
```diff
@@ -817,12 +817,14 @@ to LLVM ops. At the moment, SPIR-V module attributes are ignored.
 
 ## `mlir-spirv-cpu-runner`
 
-`mlir-spirv-cpu-runner` allows to execute `gpu` dialect kernel on the CPU via
-SPIR-V to LLVM dialect conversion. Currently, only single-threaded kernel is
+`mlir-spirv-cpu-runner` is the name of a formerly separate tool that allows to
```
I would rename it to something like: 'SPIR-V CPU Runner` and omit the old tool in the description entirely -- I think it should be enough to explain how to run things in the flow

andfau-amd replied:
Yeah that's a good idea. I've done this now, hope the new version looks good.

kuhar reviewed (changes requested):
I agree with @joker-eph, the link stage should be done at the mlir-opt level, unless we have a very good reason for linking llvm IR over MLIR

rengolin commented:
Can we just call it `mlir-runner`, please?

andfau-amd commented:
> Can we just call it `mlir-runner`, please?

I do think this would be a sensible change, but I think it should be done once all the separate runners are gone, as its own commit.

kuhar reviewed (approved):
LGTM % the documentation.

But let's also wait for a confirmation from @joker-eph.

andfau-amd committed:
[mlir] Remove the mlir-spirv-cpu-runner (move to mlir-cpu-runner)

This commit builds on and completes the work done in
9f6c632ecda08bfff76b798c46d5d7cfde57b5e9 to eliminate the need for a
separate mlir-spirv-cpu-runner binary. Since the MLIR processing is
already done outside this runner, the only real difference between it
and the mlir-cpu-runner is the final linking step between the nested
LLVM IR modules. By moving this step into mlir-cpu-runner behind a new
command-line flag (`--link-nested-modules`), this commit is able to
completely remove the runner component of the mlir-spirv-cpu-runner.

The runtime libraries and the tests are moved and renamed to fit into
the Execution Engine and Integration tests, following the model of the
similar migration done for the CUDA Runner in D97463.

andfau-amd commented on mlir/tools/mlir-cpu-runner/mlir-cpu-runner.cpp:
```diff
@@ -17,10 +17,68 @@
 
+using namespace mlir;
+
+// TODO: Consider removing this linking functionality from the SPIR-V CPU Runner
+//       flow in favour of a more proper host/device split like other runners.
+//       https://github.com/llvm/llvm-project/issues/115348
```
Here's the new TODO (https://github.com/llvm/llvm-project/issues/115348)
